frontend/packages/console-shared/src/hooks/scroll.ts

```diff
@@ -0,0 +1,53 @@
+import * as React from 'react';
+
+export enum ScrollDirection {
+  scrollingUp = 'scrolling-up',
+  scrollingDown = 'scrolling-down',
+  scrolledToBottom = 'scrolled-to-bottom',
+  scrolledToTop = 'scrolled-to-top',
+}
+
+export const useScrollDirection = (node: HTMLElement) => {
+  const scrollPosition = React.useRef(null);
+  const [scrollDirection, setScrollDirection] = React.useState<ScrollDirection>(null);
+  const handleScroll = React.useCallback(
+    (event) => {
+      const { scrollHeight, scrollTop, clientHeight } = event.target;
+      if (!scrollPosition.current) {
+        scrollPosition.current = scrollTop;
+        return;
+      }
+      if (scrollHeight - scrollTop === clientHeight) {
+        scrollPosition.current = scrollTop;
+        setScrollDirection(ScrollDirection.scrolledToBottom);
+      }
+      if (scrollTop === 0) {
+        scrollPosition.current = scrollTop;
+        setScrollDirection(ScrollDirection.scrolledToTop);
+      }
+      if (scrollPosition.current > scrollTop) {
+        scrollPosition.current = scrollTop;
+        if (scrollDirection !== ScrollDirection.scrollingUp)
+          setScrollDirection(ScrollDirection.scrollingUp);
+      }
+      if (scrollPosition.current < scrollTop) {
+        scrollPosition.current = scrollTop;
+        if (scrollDirection !== ScrollDirection.scrollingDown)
+          setScrollDirection(ScrollDirection.scrollingDown);
+      }
```

Comment:
```suggestion
      if (scrollPosition.current) {
        if (scrollHeight - scrollTop === clientHeight) {
          setScrollDirection(ScrollDirection.scrolledToBottom);
        }
        if (scrollTop === 0) {
          setScrollDirection(ScrollDirection.scrolledToTop);
        }
        if (scrollPosition.current > scrollTop) {
          if (scrollDirection !== ScrollDirection.scrollingUp)
            setScrollDirection(ScrollDirection.scrollingUp);
        }
        if (scrollPosition.current < scrollTop) {
          if (scrollDirection !== ScrollDirection.scrollingDown)
            setScrollDirection(ScrollDirection.scrollingDown);
        }
      }
      scrollPosition.current = scrollTop;
```

Reply:
This would cause some fluctuation. when scrollTop is 0 first it would go to ScrolledToTop the it would also go to ScrollingUp because scrollPosition.current didn't change. similarly for the ScrolledToBottom will happen

Reply:
I didn't change the logic I don't think... I just wrapped everything and made the update scrollPosition be at the bottom. I think it's the exact same flow / logic, is it not?

Reply:
Tested it in isolation and discussed it with Sahil... there is a bug here that is made extra apparent by my change. Sahil will be breaking this out into a utility that returns the `ScrollDirection` and writing a tests for it. In the effect, it'll do whatever extraction it needs, and just sends simple information into a utility that returns the `ScrollDirection` to set into the local state variable.

Definitely a poster-child case for having tests around all logic :)

Reply:
Done